### PR TITLE
Add a resource registry

### PR DESCRIPTION
**Pull request by Nedelosk**
Re-opening of: #3151

### CODE REVIEW

**Comment by mezz** [comment]
Copying my response from the earlier PR https://github.com/MinecraftForge/MinecraftForge/pull/3151

> This is an improvement over the OreDictionary for these particular types of items.
> This can unify resources so that only one type is used by mods. This is different from the OreDictionary which allows many different items to act as equivalent in recipes.
> 
> I experimented with using the OreDictionary to unify ores in this way and there are some problems.
> First, OreDictionary.WILDCARD_VALUE is an annoying issue. If there is ingotCopper registered to the OreDictionary with an itemstack that has wildcard meta, that's perfectly valid for recipes. It's not valid for other purposes like block drops which are important here.
> 
> This also has a better and more specific meaning than the OreDictionary tags, these are specific processing steps for resources. Things like "nugget" are not always used for ores, and relying purely on conventions is dangerous for a mod that turns ingots into nuggets. Forestry for example wants to pick up all resources and their derivatives with its Miner's backpack, but can't reliably use the ore dictionary for that purpose.
> 
> OreDictionary is being relied on for too many things, and it is not a good specialized tool for this common purpose. New tools need to be created, and this is a step in the right direction. I think the discussion should focus on fixing up this PR instead of shooting it down.

**Comment by shadowfacts** [comment]
If this is going to be a thing, I think it should replace the OreDict all at once (so in 1.11 probably) because otherwise there will be issues with mods supporting both caused by 2 equally official standards.

As for the metadata wildcard issue, I think it's a reasonably safe assumption to make that the item with the metadata 0 is equivalent to the item with the wildcard metadata.

File: src/main/java/net/minecraftforge/resources/Resource.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+{
+    /** The unique identification name for this resource. */
+    protected final String resourceName;
+
+    /**
+     * The productions of the resource.
+     * The string is the type of production like: ingot, nugget, block, ore, fluid, gas, gem.
+     * The item stack is the production.
+     */
+    static BiMap<String, ItemStack> productions = HashBiMap.create();
```
Comment (kashike) [inline src/main/java/net/minecraftforge/resources/Resource.java]: Is this supposed to be `static` if it's for this specific `Resource`?

**Comment by RANKSHANK** [comment]
I have expressed similar reservations with the misusage/expectations of the ore dictionary in #3083

If a new system comes into place I believe a backend for quantifying a ResourceStack's material value should be packaged along with it for mods like Tinkerer's construct that intend to 'break down' forms of a given resource

**Comment by williewillus** [comment]
This has been rejected https://github.com/MinecraftForge/MinecraftForge/pull/3151#issuecomment-237057453

**Comment by asiekierka** [comment]
I like this more than the existing OreDict for metals, but it does feel limited for non-metals. Better than the Ingredient concept in my view, however...

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Pull request closed without merging**